subsys/bluetooth/host/iso.c

```diff
@@ -380,10 +400,10 @@ static struct net_buf *hci_le_set_cig_params(struct bt_iso_create_param *param)
 
 		switch (param->chans[i]->qos->dir) {
 		case BT_ISO_CHAN_QOS_IN:
-			cis->m_sdu = param->chans[i]->qos->sdu;
+			cis->s_sdu = param->chans[i]->qos->sdu;
```

Comment:
Unless QOS IN has the opposite meaning of the HCI IN/OUT this needs to be m_sdu.
From core specification:
Input (Host to Controller) - hence need a m_sdu > 0
Output (Controller to Host) - hence a s_sdu > 0

Unless this code is executed on the peripheral device?

Reply:
That's a really good point. Since this is in the host, I had always assumed that `BT_ISO_CHAN_QOS_IN` was incoming data to the host, where "IN" in terms of HCI is indeed from host to controller. 

From the `bt_iso_setup_data_path` function, we seemingly set the `in_path` (i.e. host to controller) if `BT_ISO_CHAN_QOS_IN` is set:
```
	struct bt_iso_data_path out_path = { .dir = BT_ISO_CHAN_CTRL_TO_HOST };
	struct bt_iso_data_path in_path = { .dir = BT_ISO_CHAN_HOST_TO_CTRL };
	...
	case BT_ISO_CHAN_QOS_IN:
		in_path.pid = in_path.path->pid;
		out_path.pid = BT_ISO_DATA_PATH_DISABLED;
		break;
	case BT_ISO_CHAN_QOS_OUT:
		in_path.pid = BT_ISO_DATA_PATH_DISABLED;
		out_path.pid = out_path.path->pid;
		break;
	case BT_ISO_CHAN_QOS_INOUT:
		in_path.pid = in_path.path->pid;
		out_path.pid = out_path.path->pid;
		break;
```
So based on that, BT_ISO_CHAN_QOS_IN does mean from host to controller, and BT_ISO_CHAN_QOS_OUT is from controller to host, and this change isn't correct.

Reply:
But the m_sdu/s_sdu has nothing to do with data path direction, data path direction is between host and controller, Master and Slave is between devices, so it seems it was already broken when we assuming BT_ISO_CHAN_QOS_IN means input data path direction.

Reply:
@Vudentz I think it should be made clear in the header file what `BT_ISO_CHAN_QOS_IN` and `BT_ISO_CHAN_QOS_OUT` means, because it is a bit confusing that `IN` means out and `OUT` means in. 

We could consider have the values reversed of those in HCI, i.e. IN => in and OUT => out as this is typically used from the host's perspective. 

I think that change, either adding documentation in the header file or changing IN to OUT, fits nicely in this PR as well.

Reply:
When thinking on it again, QOS settings are applied on the central side when setting the CIS parameters, but are used on both central side and peripheral side when applying the data path settings hence in relation to m_sdu and s_sdu the relation to input/output will be the opposite in the peripheral compared to the central.
But only the central side will need the m_sdu/s_sdu in this case, and I believe that it makes sense in a Bluetooth stack  to define input as data going from some subsystem (audio for example) into the Bluetooth subsystem and output as data going out of the Bluetooth subsystem. Therefore there is also a 1-to-1 mapping between in-/out and the m_sdu and s_sdu.
As we have already interchanged this code several times in the past, there is clearly a confusion about what IN and OUT means - but I se no reason to define the QOS-IN/OUT the opposite of what is already used in the Bluetooth specification as it makes sense to call something that goes into a subsystem for input - you could also argue the opposite but as that conflicts with the already defined direction perception it makes little sense to define it otherwise.
When that is said, one option would be to rename m_sdu and s_sdu to in_sdu and out_sdu and only do the correct mapping to the role in the HCI layer.

Reply:
Just FYI, I've updated my BIS PR to use IN as the broadcaster and OUT as the receiver, as that is how it's supposed to be. I also updated the CIS sample apps PR.

Reply:
> When that is said, one option would be to rename m_sdu and s_sdu to in_sdu and out_sdu and only do the correct mapping to the role in the HCI layer. That would make code maintenance a bit easier as you would not need to take the role into consideration every time the m_ s_ values are used in the code.

I don't think we can actually do that because the sdu fields are role related, again we are talking about different things here since QOS applies to SDU master/slave vs data path direction, also we do intend to split the QOS for each direction so there we could change to use direction given by the spec but I do expect still confusion if somenthing bt_iso_chan_send to use input QoS and .recv maps to output QoS, so perhaps we should just use tx/rx which should imply transmission to another device and receiving from another device.

Reply:
RX and TX sounds good - here assuming we talk BT RX/TX and not for example audio data RX/TX (I2S/UART/SPI). What I want to avoid is code that looks like:
```
case xxx_input:
  some_var = yyy_output;
```
As that will lead to many discussions as those we have here in the future :-)
The sample above is what we have if this change is committed and what we will have if we define input => receive - which is the opposite of what the core-spec defines as input. The core specification defines input as input to the Bluetooth subsystem from other sub-systems on the device - hence input => transmit. 
IMO it makes sense to stick to a single definition in all the Bluetooth code, and I still do not see a reason to make that definition the opposite of what the specification uses, and what the controller code in Zephyr will be using.
@Vudentz I'm not sure I follow why configuration of rx-pdu and tx-sdu size is unrelated to configuration of the data path? You cannot have a data path configured as input if the tx_sdu size is 0 - hence there is a 1-to-1 relation between QoS and data path.

Reply:
We would be incosistent with the likes of L2CAP APIs, etc, if we change input to be TX, and there also the fact that the likes of btmon, etc, where we also have the same analogy of > (input) for rx and < (output) which dates back way before the introduction of data path direction.

Reply:
In bt_iso_setup_data_path() we have:
`	struct bt_iso_data_path in_path = {
		.dir = BT_HCI_DATAPATH_DIR_HOST_TO_CTLR };
...
	case BT_ISO_CHAN_QOS_IN:
		in_path.pid = in_path.path->pid;
		out_path.pid = BT_ISO_DATA_PATH_DISABLED;
		break;
`
Hence we map QOS_IN to in_path which is input according to the Bluetooth core specification (Host to controller - or in general input to the Bluetooth sub system from another subsystem on the device.) In other words the central device will need to transmit ISO data.
With this m_sdu change when selecting QOS_IN on the central side you will get a m_sdu of 0 but at the same time request the audio subsystem to generate data to be transmitted by the central - which won't work.
Hence if it is decided to redefine what input means in the host stack, we need to reverse the meaning when issuing the HCI commands.

Reply:
@Casper-Bonde-Bose BT_ISO_CHAN_QOS_IN is gone:

```
@@ -60,45 +60,43 @@ struct bt_iso_chan {
        struct bt_iso_chan_ops          *ops;
        /** Channel QoS reference */
        struct bt_iso_chan_qos          *qos;
-       /** Channel data path reference*/
-       struct bt_iso_chan_path         *path;
        sys_snode_t                     node;
        uint8_t                         state;
        bt_security_t                   required_sec_level;
 };
 
-/** @brief Audio QoS direction */
-enum {
-       BT_ISO_CHAN_QOS_IN,
-       BT_ISO_CHAN_QOS_OUT,
-       BT_ISO_CHAN_QOS_INOUT
+/** @brief ISO Channel IO QoS structure. */
+struct bt_iso_chan_io_qos {
+       /** Channel interval in us. Value range 0x0000FF - 0x0FFFFFF. */
+       uint32_t                        interval;
+       /** Channel Latency in ms. Value range 0x0005 - 0x0FA0. */
+       uint16_t                        latency;
+       /** Channel SDU. Value range 0x0000 0 0x0FFF. */
+       uint8_t                         sdu;
+       /** Channel PHY - See BT_GAP_LE_PHY for values.
+        *  Setting BT_GAP_LE_PHY_NONE disables data path.
+        */
+       uint8_t                         phy;
+       /** Channel Retransmission Number. Value range 0x00 - 0x0F. */
+       uint8_t                         rtn;
+       /** Channel data path reference.
+        *  Setting to NULL default to HCI data path.
+        */
+       struct bt_iso_chan_path         *path;
 };
 
 /** @brief ISO Channel QoS structure. */
 struct bt_iso_chan_qos {
-       /** @brief Channel direction
-        *
-        *  Possible values: BT_ISO_CHAN_QOS_IN, BT_ISO_CHAN_QOS_OUT or
-        *  BT_ISO_CHAN_QOS_INOUT. Shall be BT_ISO_CHAN_QOS_IN for broadcast
-        *  transmitting, and BT_ISO_CHAN_QOS_OUT for broadcast receiver.
-        */
-       uint8_t                         dir;
-       /** Channel interval in us. Value range 0x0000FF - 0x0FFFFFF. */
-       uint32_t                        interval;
        /** Channel SCA - Only for CIS */
        uint8_t                         sca;
        /** Channel packing mode. 0 for unpacked, 1 for packed. */
        uint8_t                         packing;
        /** Channel framing mode. 0 for unframed, 1 for framed. */
        uint8_t                         framing;
-       /** Channel Latency in ms. Value range 0x0005 - 0x0FA0. */
-       uint16_t                        latency;
-       /** Channel SDU. Value range 0x0000 0 0x0FFF. */
-       uint8_t                         sdu;
-       /** Channel PHY - See BT_GAP_LE_PHY for values. Shall not be BT_GAP_LE_PHY_NONE. */
-       uint8_t                         phy;
-       /** Channel Retransmission Number. Value range 0x00 - 0x0F. */
-       uint8_t                         rtn;
+       /* Channel Receiving QoS: Data Path BT_HCI_DATAPATH_DIR_CTLR_TO_HOST */
+       struct bt_iso_chan_io_qos       rx;
+       /* Channel Transmission QoS: Data Path BT_HCI_DATAPATH_DIR_CTLR_TO_HOST */
+       struct bt_iso_chan_io_qos       tx;
 };

```

Reply:
@Vudentz This makes much more sense, thanks!